docs/features/incremental-generators.cookbook.md

```diff
@@ -0,0 +1,810 @@
+    <AdditionalFiles Include="file3.txt" MyGenerator_EnableLogging="false" /> <!-- never enable logging for this file -->
+</ItemGroup>
+```
+
+**Full Example:**
+
+MyGenerator.props:
+
+```xml
+<Project>
+    <ItemGroup>
+        <CompilerVisibleProperty Include="MyGenerator_EnableLogging" />
+        <CompilerVisibleItemMetadata Include="AdditionalFiles" MetadataName="MyGenerator_EnableLogging" />
+    </ItemGroup>
+</Project>
+```
+
+MyGenerator.csproj:
+
+```xml
+<Project>
+  <PropertyGroup>
+    <GeneratePackageOnBuild>true</GeneratePackageOnBuild> <!-- Generates a package at build -->
+    <IncludeBuildOutput>false</IncludeBuildOutput> <!-- Do not include the generator as a lib dependency -->
+  </PropertyGroup>
+
+  <ItemGroup>
+    <!-- Package the generator in the analyzer directory of the nuget package -->
+    <None Include="$(OutputPath)\$(AssemblyName).dll" Pack="true" PackagePath="analyzers/dotnet/cs" Visible="false" />
+
+    <!-- Package the props file -->
+    <None Include="MyGenerator.props" Pack="true" PackagePath="build" Visible="false" />
+  </ItemGroup>
+</Project>
+```
+
+MyGenerator.cs:
+
+```csharp
+
+[Generator]
+public class MyGenerator : IIncrementalSourceGenerator
+{
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        var emitLoggingPipeline = context.AdditionalFilesProvider
+            .Combine(context.AnalyzerConfigOptionsProvider)
+            .Select((pair, ctx) =>
+                pair.Right.GetOptions(pair.Left).TryGetValue("build_metadata.AdditionalFiles.MyGenerator_EnableLogging", out var perFileLoggingSwitch)
+                ? perFileLoggingSwitch
+                : pair.Right.GlobalOptions.TryGetValue("build_property.MyGenerator_EnableLogging", out var emitLoggingSwitch)
+                  ? emitLoggingSwitch
+                  : false);
+
+        var sourcePipeline = context.AdditionalOptionsProvider.Select((file, ctx) => /* Gather build info */);
+
+        context.RegisterSourceOutput(sourcePipeline.Combine(emitLoggingPipeline), (context, pair) => /* Add source */);
+    }
+}
+```
+
+### Unit Testing of Generators
+
+**User scenario**: As a generator author, I want to be able to unit test my generators to make development easier and ensure correctness.
+
+**Solution A**:
+
+The recommended approach is to use [Microsoft.CodeAnalysis.Testing](https://github.com/dotnet/roslyn-sdk/tree/main/src/Microsoft.CodeAnalysis.Testing#microsoftcodeanalysistesting) packages:
+
+- `Microsoft.CodeAnalysis.CSharp.SourceGenerators.Testing.MSTest`
+- `Microsoft.CodeAnalysis.VisualBasic.SourceGenerators.Testing.MSTest`
+- `Microsoft.CodeAnalysis.CSharp.SourceGenerators.Testing.NUnit`
+- `Microsoft.CodeAnalysis.VisualBasic.SourceGenerators.Testing.NUnit`
+- `Microsoft.CodeAnalysis.CSharp.SourceGenerators.Testing.XUnit`
+- `Microsoft.CodeAnalysis.VisualBasic.SourceGenerators.Testing.XUnit`
+
+PROTOTYPE: Write more here
```

Comment:
Did you want to address this before merging?

Reply:
I think I need help from @chsienki and @sharwell to complete this section. I don't know the current approach.

Reply:
Lets just open an issue to write these parts and remove the PROTOTYPE.